assert: enforce type check in deepStrictEqual

joyeecheung opened the pull request:
For a write-up on this PR, see https://github.com/nodejs/node/pull/10282#issuecomment-274590846

The less-breaking part of this PR has been moved to https://github.com/nodejs/node/pull/11128

<!--
Thank you for your pull request. Please review below requirements.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test nosign` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows commit guidelines

##### Affected core subsystem(s)
<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

assert, test

##### Description of change
<!-- Provide a description of the change below this comment. -->

Add checks for the built-in tags and refactor _deepEqual,
most user-land modules do this for a strict, deep comparison.
https://github.com/nodejs/node-v0.x-archive/pull/7178 already added
this check for argument objects, even in non-strict tests.

Fixes #10258.
Docs updates are postponed due to https://github.com/nodejs/node/pull/7815.

See https://tc39.github.io/ecma262/#sec-object.prototype.tostring
for a partial list of the tags.

addaleax commented on lib/assert.js:
```diff
@@ -140,90 +140,149 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+  return a.source === b.source &&
+         a.global === b.global &&
+         a.multiline === b.multiline &&
+         a.lastIndex === b.lastIndex &&
```
So, this is not related to the changes you are making here, but I would have a couple of suggestions:

- I guess you could probably make arguments both for including `lastIndex` here and for leaving it out, and I’m tempted to say “drop it”. I probably wouldn’t expect this.

- Maybe instead of the `global`, `multiline` and `ignoreCase` checks a single check for `.flags` would be a good idea? There are new flags like `u` that would currently be ignored.

(If you think these suggestions make sense, it’s probably best to keep that for a separate commit/separate PR.)

joyeecheung replied:
Good suggestion, I think I will just submit another commit in this PR.

addaleax commented on lib/assert.js:
```diff
@@ -140,90 +140,149 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+  if (ArrayBuffer.isView(actual) && ArrayBuffer.isView(expected) &&
+              getTag(actual) === getTag(expected) &&
+              !(actual instanceof Float32Array ||
+                actual instanceof Float64Array)) {
```
The indentation seems off here?

Also, could you update the `instanceof` checks here to use `getTag`, too? (probably also best kept for a separate commit or PR)

addaleax commented on test/parallel/test-assert-strict-typecheck.js:
```diff
@@ -0,0 +1,93 @@
+
+const date = new Date('2016');
+
+// See #10258
```
nit: Can I suggest that you update this to use the full URL here?

addaleax commented on test/parallel/test-assert-strict-typecheck.js:
```diff
@@ -0,0 +1,93 @@
+const fake = new FakeDate();
+
+/* eslint-disable no-restricted-properties */
+// We are testing deepEqual!
```
Is the eslint-disable tag to that `deepEqual` can be used here?

joyeecheung replied:
`no-restricted-properties` is the one I get from `make lint`, I can't find a specific eslint-disable tag for `deepEqual`, do you know what that is?

addaleax replied:
@joyeecheung Nope, and to be clear, I wasn’t requesting you to change anything here, just making sure that I understood why the line is there in the first place

joyeecheung replied:
Oh yes, this disable tag is for all thee `deepEqual`s below :).

Trott replied:
I think it's fine the way you did it, but if you want to get even more granular with the linting options, you can disable the check on just the lines that need it by either using `// eslint-disable-line no-restricted-properties` or `// eslint-disable-next-line no-restricted-properties` (depending on if you can fit the comment on the same line or if it needs to go on the line right before the `assert.deepEqual()`).

joyeecheung replied:
Thank you for the tips!

addaleax commented on test/parallel/test-assert-strict-typecheck.js:
```diff
@@ -0,0 +1,93 @@
+// For deepStrictEqual we check the runtime type,
+// then reveal the fakeness of the fake date
+assert.throws(() => assert.deepStrictEqual(date, fake),
+                    `AssertionError: ${date} deepStrictEqual Date { }`);
```
The indentation is off here, and there’s a `assert.throws` pitfall: You can’t specify a string to check the error message against; only a constructor, `RegExp` or function that checks the argument work here. (You can probably just wrap the expression in `new RegExp` or something)

joyeecheung replied:
Sorry for not reading the docs more thoroughly. I'll wrap them up in `new RegExp`, thanks!

addaleax replied:
> Sorry for not reading the docs more thoroughly.

@joyeecheung If it helps… that mistake is something found a *lot* in our own tests, at least until very recently ;)

addaleax commented on test/parallel/test-assert-strict-typecheck.js:
```diff
@@ -0,0 +1,93 @@
+const date2 = new MyDate('2016');
+
+// deepEqual return true as long as the time are the same,
+// but deepStrictEqual check properties
```
typos: `returns` and `checks` (here and below)

joyeecheung commented:
@addaleax  Thanks for the review, I've updated this PR, PTAL.

addaleax commented:
Thanks! This looks pretty good so far, I’ll have time for a full review of this later.

CI: https://ci.nodejs.org/job/node-test-commit/6668/
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/481/

Trott commented on test/parallel/test-assert.js:
```diff
@@ -96,13 +96,6 @@ assert.throws(makeBlock(a.deepEqual, /a/i, /a/));
-  const re1 = /a/;
-  re1.lastIndex = 3;
-  assert.throws(makeBlock(a.deepEqual, re1, /a/));
-}
```
Instead of removing this test, would it make sense to change it to `assert.doesNotThrow`?

Trott commented on test/parallel/test-assert.js:
```diff
@@ -215,12 +208,6 @@ assert.throws(makeBlock(a.deepStrictEqual, /a/i, /a/));
-{
-  const re1 = /a/;
-  re1.lastIndex = 3;
-  assert.throws(makeBlock(a.deepStrictEqual, re1, /a/));
```
Same here: `assert.doesNotThrow()`?

joyeecheung commented on lib/assert.js:
```diff
@@ -140,90 +140,148 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
-  // equivalent if it is also a Date object that refers to the same time.
-  } else if (util.isDate(actual) && util.isDate(expected)) {
-    return actual.getTime() === expected.getTime();
+  // 7.2. If both values are Date objects,
```
Didn't know this is a thing :O I will remove those numbers. Thanks.

joyeecheung commented:
@Trott Thank you for the review, I have updated this PR, PTAL.

Trott commented:
@nodejs/ctc Would we need to "unlock" the assert module to land this? I think the answer is: "technically yes" but there might be some wiggle room on something like this?

Trott commented:
Added the `blocked` label until the "Locked API" issue is sorted out.

joyeecheung commented:
> Added the blocked label until the "Locked API" issue is sorted out.

Just linking the issue here: https://github.com/nodejs/node/issues/6528

Trott commented:
@nodejs/ctc: Thoughts on the semver-ness of this change? And thoughts on whether we'd need to change the API status from Locked if we want to land this? 

(Long-winded aside: Personally, I think we should remove "Locked" from the `assert` API. Either we should fully deprecate everything in it except `assert()`, or else we should support it. And I vote for supporting it. But deprecation would be better than "Locked". At least for the `assert` module, "Locked" is this really problematic middle state where we're not "deprecating" it but we are discouraging people from using it, which is another word for "deprecating". It also discourages people from spending time fixing bugs in the `assert` module. Either we should deprecate it or we should support it. For the `assert` module, at least, "Locked" seems like we're just trying to have it both ways.)

addaleax commented:
> And thoughts on whether we'd need to change the API status from Locked if we want to land this?

Fwiw I think `semver-major` changes and bugfixes are not mutually exclusive terms, and I think this change here might fall under both categories; so: no, not necessarily.

jasnell commented:
Definitely semver-major

Trott commented:
Removing the `blocked` label.

italoacasas commented:
ping @joyeecheung

Trott commented:
> ping @joyeecheung

@italoacasas Are you pinging her for a rebase or something else?

I don't think this is likely to land without @nodejs/ctc review so I'm going to add the `ctc-agenda` label to this one in the meantime.

italoacasas commented:
o ok, good to know, yes I was pining because of the conflict.

joyeecheung commented:
Ah sorry, been wanting to rebase this since https://github.com/nodejs/node/pull/10579 got merged.

This is now breaking by tests introduced in https://github.com/nodejs/node/pull/10653 and https://github.com/nodejs/node/pull/9683 because these two use `deepStrictEqual` to test equality between a Uint8Array and a Buffer with the same content. I'll spend some time to investigate what the desired behavior should be.

joyeecheung commented:
Hmm, so turns out this PR fixes a subtle code path and reveals something not checked the way as the doc says:

```
const arr = new Uint8Array([120, 121, 122, 10]);
const buf = Buffer.from(arr);
```

These two has different prototypes:

```
> Object.getPrototypeOf(arr)
TypedArray {BYTES_PER_ELEMENT: 1}
> Object.getPrototypeOf(buf)
Uint8Array {}
```

Before this PR their prototypes are not properly checked. As soon as `deepStrictEqual` finds out they are both TypedArray-ish thingy and `buffer.compare` returns 0 when they are both converted to Buffers, it returns true without checking their prototypes, even thought the doc says it will check both their prototypes and their contents.

I think probably it's the buffer module that needs to be fixed in this case.

joyeecheung commented:
> I think probably it's the buffer module that needs to be fixed in this case.

Ah wait. I think it's kinda impossible for a Uint8Array and a Buffer to have the same prototype if we keep the way Buffers are implemented. :/. Right now it's more pratical to special case for buffers and don't check the prototypes, although that would contradict what the doc says.

Related issue: https://github.com/nodejs/node/issues/9531

addaleax commented on lib/assert.js:
```diff
@@ -140,95 +140,151 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+    } else if (!strict) {
+      return true;  // Skip further checks for non-strict comparison.
+    }
+  }
```
Is this something that can be removed given the `ArrayBuffer.isView` check below?

joyeecheung replied:
This would avoid two extra `Buffer.from` calls when both of them are Buffers(might be expensive). I think it is worth keeping, unless we change how TypedArrays are compared(e.g. just loop over the elements).

No sure why the TypedArray check is implemented that way in the first place though, if it's for performance then wouldn't the overhead of calling into c++ and the copying in `Buffer.from` just kill that?

addaleax replied:
> the copying in `Buffer.from` just kill that?

It’s not copying; the `Buffer.from(arrayBuffer, offset, length)` overload returns a `Buffer` whose underlying memory is contained in the passed `arrayBuffer`. So I don’t think these really count as expensive. ;)

As far as the call into C++ goes… it was benchmarked for https://github.com/nodejs/node/pull/4330, and it seems plausible to me that it’s faster than a plain JS loop (unless a lot has changed in V8 since then?).

joyeecheung replied:
Ah then I think it's OK to remove these! Thanks for the info.

addaleax commented:
> This is now breaking by tests introduced in #10653

oops 😳 I think you can feel free to fix up those tests as part of this PR.

joyeecheung commented:
Removed the buffer checks in favor of typed array checks per @addaleax 's suggestion. Also moved the primitive checks forward because their cases should be more common than those type-specific checks.

Also I think `util.isDate`, `util.isRegExp` and `isArguments` can simply be replaced by checking our cached type tags(that's essentially what happens in C++ land for `util.isDate` and `util.isRegExp` without a cached tag, if I understand correctly). But this PR is already quite heavy so these can be done later.

joyeecheung commented:
A write-up on this PR:

1. [The really breaking bit](https://github.com/joyeecheung/node/blob/a6f57e4f50c9648c10615d0a38ead517cac7f729/lib/assert.js#L200) only applies to strict comparison:

    ```js
      if (actualTag !== expectedTag) {
        return false;
      }
    ```
    which check if the results of `Object.prototype.toString.call()` on `actual` and `expect` are the same. This is to fix something like https://github.com/nodejs/node/issues/10258 and to match behaviors of user-land assertion libraries (see https://github.com/nodejs/node/issues/10258#issuecomment-266950333). The tags (i.e. internal property `[[Class]]` in ECMAScript) are also used by stuff like `util.isDate` and `util.isRegExp` from the C++ side and is customizable from JS land since ES2015 via `Symbol.toStringTag`.

    In a sense this check makes `deepStrictEqual` more strict, which could be desirable for tests and validations (probably this is why user land modules do this check), and separates it further apart from `deepEqual`, which doesn't check any types conceptually.

    Another reason for this check is that this is already introduced in https://github.com/nodejs/node-v0.x-archive/pull/7178 for `arguments` objects, and in that case it applies to strict and non-strict comparison alike.
2. This PR refactors the `_deepEqual` function a bit and fixes a few code paths that lead to behaviors contradicting to what the doc says (e.g. https://github.com/nodejs/node/issues/10258#issuecomment-266963234 and https://github.com/nodejs/node/pull/10282#issuecomment-274267895). Before this PR certain types of objects(Buffers, Dates, etc.) are not checked properly, and can get away with different prototypes AND different enumerable owned properties because `_deepEqual` would jump to premature conclusion for them. 

1 fixes what some people might consider as bugs (especially when they are switching from user land modules to built-in `assert`, which is a trend encouraged by something like [power-aasert](https://github.com/power-assert-js/power-assert)), and 2 fixes what are definitely bugs. 2 doesn't rely on 1 though, if 1 is not desired we can simply just remove that three lines of code above.

joyeecheung commented:
Also, we could reduce the check of tags to just a set of built-in objects(probably just the ones specified in https://tc39.github.io/ecma262/#sec-object.prototype.tostring), instead of checking all objects, as the tag is customizable.

ChALkeR commented:
Tangentially related: #6528.

addaleax commented on lib/assert.js:
```diff
@@ -140,95 +139,149 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+      return false;
+    } else if (!strict) {
+      return true;  // Skip further checks for non-strict comparison.
+    }
```
Just noticed… for non-strict comparison, when the `areSimilarTypedArrays` check returns true, this continues and checks all own properties of the buffers anyway (including the indexed ones), right? That probably undoes the performance gains for which this optimization was originally intended for? :/

joyeecheung replied:
Hmm..but that optimization would allow such cases:

```js
var a = Buffer.from([1, 2, 3]);
var b = Buffer.from([1, 2, 3]);
b.prop = 1;
assert.deepStrictEqual(a, b);  // doesn't throw
```

I think one way to reduce the performance hit is to filter out numeric keys for Buffers before checking them?

addaleax replied:
Yeah, I’m not sure what the best way is either :/ Filtering out numeric keys seems good but we’d probably want to benchmark that kind of thing?

joyeecheung replied:
I'm not sure what this should be benchmarked against though, the original version would be significantly faster because it allows false positive, which is what this one tries to fix?

(Aside: I noticed this is what repl does for completion too)

addaleax replied:
Sorry – let me clarify a bit: I meant “benchmark” in the sense that it might be good to have a feeling for how big the performance differences are, not that I’d want to make landing this PR dependent on its performance implications.

joyeecheung replied:
Ah I understand. I'll post some numbers later :D

ljharb commented:
@joyeecheung the [[Class]] internal slot in ES5 was removed in ES6, and Object.prototype.toString output is *no longer reliable* for brand-checking due to Symbol.toStringTag. The actual checks required include a try/catch on a built in method that does the brand check for you - see npmjs.com/is-string, npmjs.com/is-number-object, and many more of my "is-" packages.

I would strongly encourage node core to use the reliable approach instead of the deceptively easy one.

addaleax commented:
I would not actually be too worried about that (or rather, it depends on what you mean by reliable)… if somebody sets the @@toStringTag on an object to `Uint8Array`, they probably have a reason for that, and probably even want libraries to treat it like an actual `Uint8Array`?

ljharb commented:
That's certainly one interpretation, but because it lacks the internet slots, it *isnt* an array, and built in methods will treat it differently. As such, if you want it to be like a Uint8Array, the proper way to do that is `class extends Uint8Array` - which will provide the internal slots and make the checks I'm advocating pass.

In other words, toStringTag is only useful to pretend to be an object that lacks internal slots - which doesn't apply to any builtin.

joyeecheung commented:
@ljharb Thanks for the pointers! I took a look into the spec and the modules, and I think `assert`  already tries to do something similar to the `is-` packages before this PR, though the type-specific checks in `assert` might be different from what the `is-` packages choose (could be a CommonJS legacy, but since we've removed the Common JS spec numbers from the code, I think this can be addressed in a separate PR updating those selected checks).

If I understand correctly, this PR actually makes the check stricter than the is- packages. Prior to this PR, `assert` only does the type-specific checks, like what those packages do once they detect that the tags can be tampered with. After this PR, it checks the tags first, but even when the arguments do have the same tag, it still proceeds with type-specific checks. The tag check only provides an early exit point for `false`, but not for `true`.

Also this probably should be a `util` issue? Some of the `util.is*` methods rely on `v8::Value::Is*` and AFAICT those would check some V8 internal types, which can be used by V8 to implement the type-specific built-in methods anyway, so I don't really know what needs to be fixed without checking out those implementations?

joyeecheung commented:
@ljharb Thanks for the pointers! I took a look into the spec and the modules, and I think `assert`  already tries to do something similar to the `is-` packages before this PR, though the type-specific checks in `assert` might be different from what the `is-` packages choose (could be a CommonJS legacy, but since we've removed the Common JS spec numbers from the code, I think this can be addressed in a separate PR updating those selected checks).

If I understand correctly, this PR actually makes the check stricter than the is- packages. Prior to this PR, `assert` only does the type-specific checks, like what those packages do once they detect that the tags can be tampered with. After this PR, it checks the tags first, but even when the arguments do have the same tag, it still proceeds with type-specific checks. The tag check only provides an early exit point for `false`, but not for `true`.

Also this probably should be a `util` issue? Some of the `util.is*` methods rely on `v8::Value::Is*` and AFAICT those would check some V8 internal types, which can be used by V8 to implement the type-specific built-in methods anyway, so I don't really know what needs to be fixed without checking out those implementations?

joyeecheung commented:
@ljharb Thanks for the pointers! I took a look into the spec and the modules, and I think `assert`  already tries to do something similar to the `is-` packages before this PR, though the type-specific checks in `assert` might be different from what the `is-` packages choose (could be a CommonJS legacy, but since we've removed the Common JS spec numbers from the code, I think this can be addressed in a separate PR updating those selected checks).

If I understand correctly, this PR actually makes the check stricter than the is- packages. Prior to this PR, `assert` only does the type-specific checks, like what those packages do once they detect that the tags can be tampered with. After this PR, it checks the tags first, but even when the arguments do have the same tag, it still proceeds with type-specific checks. The tag check only provides an early exit point for `false`, but not for `true`.

Also this probably should be a `util` issue? Some of the `util.is*` methods rely on `v8::Value::Is*` and AFAICT those would check some V8 internal types, which can be used by V8 to implement the type-specific built-in methods anyway, so I don't really know what needs to be fixed without checking out those implementations?

joyeecheung commented:
Ooops, something went wrong with my Internet connection and I sent out the comments multiple times :/. Sorry, I've deleted the duplicates.

joyeecheung commented:
Also I am not very sure about how strict the type check should be, given that tags can be tampered with via `@@toStringTag`. Should it

1. Demand `actual` and `expected` have the same tag, even if they pass the type-specific checks (i.e. are the same type of ducks) (this is the strictest one and is what this PR currently does, but even user-land modules aren't this strict)
2. Only check the tags of `actual` and `expected` to  when they are [built-in objects with pre-defiend tags](https://tc39.github.io/ecma262/#sec-object.prototype.tostring) (kinda what user-land modules do at the moment).
    1. If *one of* them isn't, we just check if they are the same type of ducks (doesn't fix https://github.com/nodejs/node/issues/10258)
    2. If *both* of them aren't, do the duck type check (would fix https://github.com/nodejs/node/issues/10258)
3. Or just forget about the tags since they can be tampered with anyway, just do the duck type check (what @ljharb 's is- packages do, and is the loosest of all)

ljharb commented:
@joyeecheung I wouldn't say it's looser - the tags have zero correlation to what a thing *is* because they're simply a mutable label for what a thing *claims* to be. The strictest thing is to ignore them entirely (in an environment where `Symbol.toStringTag` exists, at least).

joyeecheung commented:
@ljharb Ah I mean it's stricter in the sense like `hasSameTag && areDucks` v.s. `areDucks`

ljharb commented:
Gotcha - although in this case it's not duck typing, it's brand checking - which is the strictest check you can do. So it's more like "hasDuckDNA" than "quacksLikeDucks" :-p

joyeecheung commented:
Done some benchmarking work per @addaleax's suggestion. Opened another PR for benchmarks (https://github.com/nodejs/node/pull/11092) because I don't feel those should be blocked by this.. 

Also reorganized the commits and separated the two different parts described in https://github.com/nodejs/node/pull/10282#issuecomment-274590846. The first commit being the not-so-breaking refactor+bugfix patch, the second being the tag check.

I've run the benchmarks with some smaller setups and observed about ~90% slow down in buffer deep strict benchmarks, ~50% slow down in deep strict typed array benchmarks. Other benchmarks (don't involve typedarray/buffers or are deepEqual benchmarks) don't show significant impacts. Primitive comparisons even have performance improvements(vary from 1%-8%). Will post more numbers when I've found the time to run the whole thing..

joyeecheung commented:
CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/559/
CI: https://ci.nodejs.org/job/node-test-pull-request/6146/

joyeecheung commented:
Benchmarks are taken from https://github.com/nodejs/node/issues/11092, run against this PR'second commit and the master. The results changed after I updated the benchmarks to address the reviews though..

<details><summary>Results</summary>
<pre>
                                                                                                  improvement confidence      p.value
 assert/deepequal-buffer.js method="nonstrict" len=100 n=1000                                        -92.15 %        *** 5.592045e-07
 assert/deepequal-buffer.js method="strict" len=100 n=1000                                           -87.91 %        *** 1.481929e-10
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="array"          1.54 %            3.813086e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="boolean"        0.08 %            9.540203e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="new-array"      2.27 %            5.220313e-02
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="null"           0.26 %            8.642667e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="number"         0.73 %            4.837192e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="object"        -0.27 %            7.277261e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="string"        -0.78 %            3.996062e-01
 assert/deepequal-prims-and-objs-big-array.js method="nonstrict" len=100000 n=25 prim="undefined"      3.12 %          * 3.893927e-02
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="array"            -2.53 %            5.028944e-02
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="boolean"          -3.07 %         ** 6.539822e-03
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="new-array"        -1.14 %            3.776449e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="null"             -3.47 %            5.452339e-02
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="number"            0.30 %            8.308053e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="object"           -1.31 %          * 4.744264e-02
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="string"           -1.51 %            1.445681e-01
 assert/deepequal-prims-and-objs-big-array.js method="strict" len=100000 n=25 prim="undefined"         0.61 %            8.530695e-01
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="array"                 2.80 %            2.401182e-01
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="boolean"               7.29 %         ** 3.027695e-03
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="new-array"             3.89 %            1.033384e-01
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="null"                  6.39 %          * 4.844864e-02
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="number"                8.26 %          * 3.295566e-02
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="object"                7.04 %          * 1.246946e-02
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="string"                7.29 %          * 4.308442e-02
 assert/deepequal-prims-and-objs-big-loop.js method="nonstrict" n=1000000 prim="undefined"             5.05 %            8.801488e-02
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="array"                    7.63 %         ** 6.295852e-03
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="boolean"                  9.38 %        *** 3.859246e-04
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="new-array"                5.19 %            9.868693e-02
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="null"                     6.96 %         ** 3.754126e-03
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="number"                   8.75 %         ** 9.927209e-03
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="object"                   9.90 %        *** 8.087247e-04
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="string"                   7.00 %         ** 7.527566e-03
 assert/deepequal-prims-and-objs-big-loop.js method="strict" n=1000000 prim="undefined"               12.38 %        *** 8.321948e-06
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Float32Array"                0.05 %            9.780129e-01
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Float64Array"                1.63 %          * 3.978530e-02
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Int16Array"                -98.85 %        *** 4.075302e-10
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Int32Array"                -97.98 %        *** 5.272991e-11
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Int8Array"                 -99.33 %        *** 1.400483e-12
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint16Array"               -98.92 %        *** 8.256519e-11
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint32Array"               -98.05 %        *** 3.837230e-12
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint8Array"                -99.30 %        *** 1.832656e-09
 assert/deepequal-typedarrays.js len=1000000 method="nonstrict" n=1 type="Uint8ClampedArray"         -99.29 %        *** 1.861473e-09
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Float32Array"                  -1.03 %            4.781549e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Float64Array"                   1.46 %            1.871156e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Int16Array"                    -8.69 %          * 2.676116e-02
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Int32Array"                    -2.51 %            5.762043e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Int8Array"                     -8.23 %            1.700721e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint16Array"                   -3.40 %            3.097808e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint32Array"                    1.62 %            6.970420e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint8Array"                    -1.38 %            7.340172e-01
 assert/deepequal-typedarrays.js len=1000000 method="strict" n=1 type="Uint8ClampedArray"             -9.24 %            5.709597e-02
</pre>
</details>

joyeecheung commented:
After reading the meeting notes I think we can land the first commit for a start, this only refactors the code and fixes the premature conclusion bugs, and doesn't need to change the docs. I can open another PR for the second commit, which is more intrusive and needs to introduce a doc change about the tag checks? @Trott @addaleax @jasnell

jasnell commented:
SGTM

Trott commented:
I think that works for me.

TimothyGu commented on lib/assert.js:
```diff
@@ -140,95 +139,154 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
-        return true;
-      }
+  // Notes: Type tags are [[Class]] properties returned by
+  // obj->class_name() in C++ or Object.prototype.toString.call(obj) in JS
```
`[[Class]]` is gone in ES6, and `obj->class_name()` doesn't seem to be part of the public V8 API. I don't think we really need to mention them.

joyeecheung replied:
I mentioned `obj->class_name()` because that looks like it's what `util.isDate()` (`Value::IsDate`) and alike come down to in a macro. Also it's why `process` and a lot of other objects created in the C++ land has their own tags, although after a quick search looks like they actually use `FunctionTemplate::SetClassName()` to set this, and never use the getter, so probably we can just mention `FunctionTemplate::SetClassName()` would alter this.

`[[Class]]` is described as "historical" in another PR, just as it is in the spec note.

joyeecheung replied:
I mentioned `obj->class_name()` because that looks like it's what `util.isDate()` (`Value::IsDate`) and alike come down to in a macro. Also it's why `process` and a lot of other objects created in the C++ land has their own tags, although after a quick search looks like they actually use `FunctionTemplate::SetClassName()` to set this, and never use the getter, so probably we can just mention `FunctionTemplate::SetClassName()` would alter this.

`[[Class]]` is described as "historical" in another PR, just as it is in the spec note.

ljharb commented on lib/assert.js:
```diff
@@ -140,95 +139,158 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+}
+
+function isArguments(tag) {
+  return tag == '[object Arguments]';
```
why is this `==` and a few lines above, `===`? I assume this should be `===`?

joyeecheung replied:
Oops, it's mistake, thanks for catching that.

ljharb commented on lib/assert.js:
```diff
@@ -140,95 +139,158 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+  // See https://tc39.github.io/ecma262/#sec-object.prototype.tostring
+  // for a list of tags pre-defined in the spec.
+  // There are some unspecified tags in the wild too(e.g. typed array tags).
+  const actualTag = Object.prototype.toString.call(actual);
```
For the sake of robustness, please cache `Object.prototype.toString` at module level, and don't rely on nobody having monkeyed with `Object.prototype` at runtime :-)

ljharb commented on lib/assert.js:
```diff
@@ -140,95 +139,158 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+  // If both values are Date objects,
+  // check if the time underneath are equal first.
+  if (util.isDate(actual) && util.isDate(expected)) {
+    if (actual.getTime() !== expected.getTime()) {
```
It's safer to cache `Date.prototype.getTime` at module level, and `.call` it here - rather than relying both on `.getTime` not being shadowed on these two objects, and on `Date.prototype` being unbroken.

ljharb commented on lib/assert.js:
```diff
@@ -140,95 +139,158 @@ assert.deepStrictEqual = function deepStrictEqual(actual, expected, message) {
+    // Here compare returns true, so actual.length === expected.length
+    // if they both only contain numeric keys,
+    // we don't need to exam further
+    if (Object.keys(actual).length === actual.length &&
```
Similarly, `Object.keys` should be cached at module level, not runtime-looked-up on `Object`

joyeecheung commented:
This is blocked by https://github.com/nodejs/node/pull/11128

ChALkeR commented:
> Added the `blocked` label until the "Locked API" issue is sorted out. 

`assert` was unlocked in #11304, removing `blocked` label.

> This is blocked by #11128

That one landed too.

joyeecheung commented:
@ChALkeR Yeah this is not blocked anymore. Thanks!

Did a rebase locally, haven't decided about https://github.com/nodejs/node/pull/10282#issuecomment-275194045, I'll push it later.

joyeecheung commented:
Rebased with documentation and a few more test cases (for faked `global` and `process`)

addaleax commented on doc/api/assert.md:
```diff
@@ -125,12 +125,13 @@ changes:
   ( `===` ).
 2. [`[[Prototype]]`][prototype-spec] of objects are compared using
   the [Strict Equality Comparison][] too.
+3. [Type tags][Object.prototype.toString()] of objects should be the same.
```
I can’t find the definition of `[Object.prototype.toString()]` … am I missing something?

joyeecheung replied:
Uh, forgot to put the reference link below :P

addaleax commented on doc/api/assert.md:
```diff
@@ -141,6 +142,25 @@ assert.deepEqual({a: 1}, {a: '1'});
 // AssertionError: { a: 1 } deepStrictEqual { a: '1' }
 // because 1 !== '1' using strict equality
+
+// The following objects don't have owned properties
```
typo: `own properties`?

joyeecheung commented:
@addaleax Thanks for the review, fixed the typo and the missing link.

CI: https://ci.nodejs.org/job/node-test-pull-request/6804/

addaleax commented:
The last CITGM run is also a bit old, so:

CITGM: https://ci.nodejs.org/view/Node.js-citgm/job/citgm-smoker/641/

addaleax commented:
Went ahead and landed this in efec14a7d12d9e07e4efcc8e5c0ad7f80c47f83f :)
